### DISCUSSION TITLE
`eth_signTypedData_v4` method does not exist / is not available when connecting MetaMask on mobile browser, or on desktop browser via QR code

### DISCUSSION THREAD

**Discussion by SergeyVolynkin** [WalletConnect/walletconnect-monorepo#3595] [Developer Support] [answered]
Hi, could you please suggest how to fix 
`
The method does not exist / is not available. Details: the method eth_signTypedData_v4 does not exist/is not available Version: viem@1.7.0
`
error when user connects MetaMask wallet via WalletConnect on mobile browser / or on desktop browser via QR Code 

Was not able to find docs on `eth_signTypedData_v4` method availability on MetaMask connected via QR Code -or- mobile browser (Safari, iOS)

versions used:
```
    "@web3modal/ethereum": "^2.7.1",
    "@web3modal/react": "^2.7.1",
    "viem": "^1.7.0",
    "wagmi": "^1.3.10"
```

(side note: with the same config WalletConnect connects to MetaMask plugin when in desktop browser just fine, but on mobile browser or in desktop browser via QR code having issues)

**Comment by glitch-txs** [comment]
hi, could you share you configuration in the App file?

**Comment by SergeyVolynkin** [comment]
Hi @glitch-txs, the configuration is the following:

```tsx
import { ReactNode, useMemo } from 'react';
import {
  EthereumClient,
  w3mConnectors,
  w3mProvider,
} from '@web3modal/ethereum';
import {
  Chain,
  configureChains,
  createConfig,
  WagmiConfig as WagmiConfigProvider,
} from 'wagmi';
import { goerli, mainnet } from 'wagmi/chains';

// …

/**
 * @see https://walletconnect.com/explorer
 */
const RECOMMENDED_WALLET_IDS = [
  'c57ca95b47569778a828d19178114f4db188b89b763c899ba0be274e97267d96', // MetaMask
  '5864e2ced7c293ed18ac35e0db085c09ed567d67346ccb6f58a0327a75137489', // Fireblocks
];

export function WalletConnectConfigProvider(props: Props) {
  const config = useSystemConfig();
  const ethereumChainId = config.ethereum.chainId;

  const walletConnectConfig = useMemo(() => {
    const projectId = WALLET_CONNECT_PROJECT_ID;
    const recommendedWalletIds = RECOMMENDED_WALLET_IDS;
    const isMainnet = ethereumChainId === mainnet.id;
    const chain: Chain = isMainnet ? mainnet : goerli;
    const availableChains: [Chain] = [chain];

    const { publicClient } = configureChains(availableChains, [
      w3mProvider({ projectId }),
    ]);
    const wagmiConfig = createConfig({
      autoConnect: true,
      connectors: w3mConnectors({ projectId, chains: availableChains }),
      publicClient,
    });
    const ethereumClient = new EthereumClient(wagmiConfig, availableChains);

    return {
      projectId,
      chain,
      ethereumClient,
      recommendedWalletIds,
      wagmiConfig,
    };
  }, [ethereumChainId]);

  return (
    <WagmiConfigProvider config={walletConnectConfig.wagmiConfig}>
      <ConfigProvider value={walletConnectConfig}>
        {props.children}
      </ConfigProvider>
    </WagmiConfigProvider>
  );
}

```

The `Web3Modal` usage is the following:

```tsx
import { Web3Modal } from '@web3modal/react';

const theme = {
  // …
};

export default function WalletConnectModal() {
  const config = useWalletConnectConfig();

  return (
    <Web3Modal
      projectId={config.projectId}
      ethereumClient={config.ethereumClient}
      themeMode="dark"
      explorerRecommendedWalletIds={config.recommendedWalletIds}
      explorerExcludedWalletIds="ALL"
      defaultChain={config.chain}
      termsOfServiceUrl={TERMS_OF_SERVICE_URL}
      privacyPolicyUrl={PRIVACY_POLICY_URL}
      themeVariables={theme}
    />
  );
}

```

(@glitch-txs side note: with the same config WalletConnect connects to MetaMask plugin just fine, but on mobile browser or in desktop browser via QR code having issues)

**Reply by glitch-txs** [reply, marked as answer]
we're gonna to fix this soon, but for now as workaround can you try 

```ts
    const availableChains: [Chain] = [chain, chain];
```

**Reply by SergeyVolynkin** [reply]
Hey, @glitch-txs, big thanks! The suggested change solves the problem with both the QR code and the mobile iOS browser 😅
```tsx
const availableChains: [Chain, Chain] = [chain, chain];
```

**Reply by SergeyVolynkin** [reply]
Hi @glitch-txs, we're now transitioning to `v3` is it okay to do `const availableChains: [Chain] = [goerli]` in `v3` or do we still need to `const availableChains: [Chain] = [goerli, goerli]` ?

**Reply by SergeyVolynkin** [reply]
verified, @glitch-txs, on `v3` the `const availableChains = [goerli, goerli]` no longer works, it seems that having two different chains is mandatory (e.g. [goerli, mainnet])

**Reply by glitch-txs** [reply]
hi, you will need to use a custom configuration for that https://docs.walletconnect.com/web3modal/react/about
and pass the chains to the WalletConnectConnector directly

**Reply by SergeyVolynkin** [reply]
Hi @glitch-txs,

we moved to the custom setup, but unfortunately using just `[chain]` does not work for mobile MetaMask (iOS App + Safari Browser)

i.e. this does not work, error `The method does not exist / is not available. Details: the method eth_signTypedData_v4 does not exist/is not available Version: viem@1.16.5` :
```ts
// 2. Create wagmiConfig
const { chains, publicClient } = configureChains(
  [sepolia],
  [walletConnectProvider({ projectId }), publicProvider()]
)
```

And this works on the same eth_signTypedData_v4 flow for mobile MetaMask (iOS App + Safari Browser) 🤔:
```ts
// 2. Create wagmiConfig
const { chains, publicClient } = configureChains(
  [sepolia, mainnet],
  [walletConnectProvider({ projectId }), publicProvider()]
)
```

versions used are the follwoing:
```json
    "@web3modal/wagmi": "^3.1.0",
    "viem": "^1.16.6",
    "wagmi": "^1.4.5"

```

**Reply by glitch-txs** [reply]
you need to add at least too in the array, you can also do

`[mainnet, mainnet]` (in the WalletConnectConnector because configureChains will clear them out)

This is a known issue with our Wagmi connector that we'll address in the upcoming weeks

**Reply by SergeyVolynkin** [reply]
Got it, thank you, @glitch-txs!
A bit worried about using `[mainnet, mainnet]` as it sounds a bit hacky –– so, for now, we'll use `[sepolia, mainnet]` in production
and will adopt just `[mainnet]` on prod, and just `[sepolia]` on testnet after `Wagmi connector` is fixed (Thank you! 🙏)
by the way our dApp is https://walletconnect.com/explorer/paradex

**Reply by glitch-txs** [reply]
Yeah it looks ugly but there's nothing wrong with it technically. If you're going to use [testnet, mainnet] I would highly recommend flipping them like [mainnet, testnet]. The first chain of the array will be required in the session proposal and some wallets don't support testnets.

**Reply by SergeyVolynkin** [reply]
Got it, thank you very much @glitch-txs; For now will be using `[mainnet, testnet]`, then will switch to `[mainnet]`/`[testnet]` as soon as `Wagmi connector` fixed
